File: 567-permutation-in-string/567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        vector<int> chars(26);
        for (int i = 0; i < s1.size(); i++) {
            chars[s1[i] -'a']++;
        }
        int left = 0;
        int right = 0;
        
        while (right < s2.size()) {
            int r = s2[right];
            if (chars[r-'a']==0) {
                chars[s2[left]-'a']++;
                left++;
                continue;
            } else {
                chars[r-'a']--;
                if (right-left+1==s1.size()) {
                    return true;
                }
            }
            right++;
        }
        return false;
    }
};

File: 567-permutation-in-string/567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        chars = [0] * 26;
        for i in range(len(s1)):
            chars[ord(s1[i])-ord('a')] += 1
            
        left = 0
        right = 0
        
        while (right < len(s2)):
            if chars[ord(s2[right])-ord('a')] == 0:
                chars[ord(s2[left])-ord('a')] += 1
                left += 1
                continue
            else:
                chars[ord(s2[right])-ord('a')] -= 1
                if right-left+1 == len(s1):
                    return True
            right += 1
            
        return False